File: ann.hs
µ = 1

-- step 
step x = 
    if x < 0
        then -1
    else 1

-- relu
relu x = max(0,x)

-- sigmoid
sigmoid x =
    1.0/(1+exp(-x))

-- sigmoid derivative
sigmoid_derivative x =
    sigmoid(x) * (1.0 - sigmoid(x))

-- auxiliar for inputs
untuple (weight, value) =
    weight*value

-- neuron function
neuron action_function inputs bias =
    action_function ( bias + foldl (+) (map untuple inputs))

-- delta rule
delta_rule (weight, value) µ w d y x =
 (w + µ * (d - y) * x, value)

weight_correction inputs =
    map delta_rule inputs

File: ann.ml

(* relu *)
let relu x = max(0,x);;

(* sigmoid *)
let sigmoid x = 1.0 /. (1.0 +. 2.71828 ** x) ;;

(* sigmoid derivative *)
let sigmoid_derivative x =
    sigmoid x *. (1.0 -. sigmoid x) ;;

(* auxiliar function to multiply elements of the tuples *)
let untuple (n, m) = n *. m ;;

(* neuron function *)
let neuron activation_function bias inputs =
    activation_function (bias +. List.fold_left (+.) 0.0 (List.map untuple inputs));;

(* delta rule *)
(*  n -> learning rate
    t -> target output
    y -> obtained output
    x -> input*)
let delta_rule (weight, value) n t y x =
 (weight +. n *. (t -. y) *. x, value) ;;

(* weight modification *)
let learn inputs expected_result =
    List.map delta_rule inputs ;;
